Frequent Mac OS test failures

MacOS 10.12.6 Sierra, Xcode-9.0. Current encfs github master.

Often (but not always) I'm getting the following failure running `build.sh`:
```
cmake version 3.9.4

CMake suite maintained and supported by Kitware (kitware.com/cmake).
-- There is no usable FDATASYNC on Apple
-- Enabled syslog logging support
-- clang-tidy not found.
-- Using vendored TinyXML2
-- Using vendored easylogging++
-- Setting C++11
-- git Version: v1.9.2-cf0f1ce2
-- Version: 1.9.2
-- Performing Test HAVE_THREAD_SAFETY_ATTRIBUTES
-- Performing Test HAVE_THREAD_SAFETY_ATTRIBUTES
-- Performing Test HAVE_THREAD_SAFETY_ATTRIBUTES -- failed to compile
-- Performing Test HAVE_STD_REGEX
-- Performing Test HAVE_STD_REGEX
-- Performing Test HAVE_STD_REGEX -- success
-- Performing Test HAVE_GNU_POSIX_REGEX
-- Performing Test HAVE_GNU_POSIX_REGEX
-- Performing Test HAVE_GNU_POSIX_REGEX -- failed to compile
-- Performing Test HAVE_POSIX_REGEX
-- Performing Test HAVE_POSIX_REGEX
-- Performing Test HAVE_POSIX_REGEX -- success
-- Performing Test HAVE_STEADY_CLOCK
-- Performing Test HAVE_STEADY_CLOCK
-- Performing Test HAVE_STEADY_CLOCK -- success
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/ur20980/src/encfs/build
[  4%] Built target tinyxml2_static
[  9%] Built target easyloggingpp
[  9%] Built target encfssh-man
[  9%] Built target encfsctl-man
[ 80%] Built target encfs
[ 80%] Built target encfs-man
[ 85%] Built target encfs-bin
[ 90%] Built target encfsctl
[ 95%] Built target checkops
[100%] Built target makekey
[  5%] Built target gtest
[ 10%] Built target easyloggingpp
[ 15%] Built target tinyxml2_static
[ 87%] Built target encfs
[ 92%] Built target gtest_main
[100%] Built target unittests
Running tests...
Test project /Users/ur20980/src/encfs/build
    Start 1: unit
1/1 Test #1: unit .............................   Passed    0.17 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   0.20 sec
++ dirname ../integration.sh
+ cd ..
+ perl -I. -MTest::Harness -e '$$Test::Harness::debug=1; runtests @ARGV;' integration/normal.t.pl integration/reverse.t.pl
integration/normal.t.pl ... 
runTests: mode=standard sudo=off
integration/normal.t.pl ... 52/132 umount(/private/var/folders/c6/lnc_0m093ys8w16md_fm1mnxhtfnj8/T/encfs-tests-JQEr/create_remount.mnt): Resource busy -- try 'diskutil unmount'
integration/normal.t.pl ... 55/132 
#   Failed test 'encfs command returns 0'
#   at integration/normal.t.pl line 441.
integration/normal.t.pl ... 61/132 cannot opendir for /var/folders/c6/lnc_0m093ys8w16md_fm1mnxhtfnj8/T/encfs-tests-JQEr/create_remount.mnt: No such file or directory at integration/normal.t.pl line 388.
cannot remove directory for /var/folders/c6/lnc_0m093ys8w16md_fm1mnxhtfnj8/T/encfs-tests-JQEr/create_remount.mnt: Resource busy at integration/normal.t.pl line 388.
cannot remove directory for /var/folders/c6/lnc_0m093ys8w16md_fm1mnxhtfnj8/T//encfs-tests-JQEr: Directory not empty at integration/normal.t.pl line 388.

#   Failed test 'working dir removed'
#   at integration/normal.t.pl line 389.

runTests: mode=paranoia sudo=off
integration/normal.t.pl ... 129/132 # Looks like you planned 132 tests but ran 131.
# Looks like you failed 2 tests of 131 run.
integration/normal.t.pl ... Dubious, test returned 2 (wstat 512, 0x200)
Failed 3/132 subtests 
	(less 2 skipped subtests: 127 okay)
integration/reverse.t.pl .. ok     

Test Summary Report
-------------------
integration/normal.t.pl (Wstat: 512 Tests: 131 Failed: 2)
  Failed tests:  55, 63
  Non-zero exit status: 2
  Parse errors: Bad plan.  You planned 132 tests but ran 131.
Files=2, Tests=162, 31 wallclock secs ( 0.08 usr  0.02 sys + 16.57 cusr  2.20 csys = 18.87 CPU)
Result: FAIL
Failed 1/2 test programs. 2/162 subtests failed.
make[3]: *** [CMakeFiles/integration] Error 255
make[2]: *** [CMakeFiles/integration.dir/all] Error 2
make[1]: *** [CMakeFiles/integration.dir/rule] Error 2
make: *** [integration] Error 2
```

And it leaves filesystem(s) mounted:
```
. . . . .
encfs@osxfuse1 on /private/var/folders/c6/lnc_0m093ys8w16md_fm1mnxhtfnj8/T/encfs-tests-OYA7/create_remount.mnt (osxfuse, local, nodev, nosuid, synchronous, mounted by ur20980)
encfs@osxfuse2 on /private/var/folders/c6/lnc_0m093ys8w16md_fm1mnxhtfnj8/T/encfs-tests-kfZN/create_remount.mnt (osxfuse, local, nodev, nosuid, synchronous, mounted by ur20980)
```

After manually unmounting these, another run (removed `build/` and recompiled everything):
```
. . . . .
[100%] Linking CXX executable unittests
ld: warning: directory not found for option '-L/Users/ur20980/src/encfs/build/vendor/github.com/google/googletest/googletest'
[100%] Built target unittests
Running tests...
Test project /Users/ur20980/src/encfs/build
    Start 1: unit
1/1 Test #1: unit .............................   Passed    0.07 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   0.07 sec
Scanning dependencies of target integration
++ dirname ../integration.sh
+ cd ..
+ perl -I. -MTest::Harness -e '$$Test::Harness::debug=1; runtests @ARGV;' integration/normal.t.pl integration/reverse.t.pl
integration/normal.t.pl ... 
runTests: mode=standard sudo=off
integration/normal.t.pl ... 52/132 umount(/private/var/folders/c6/lnc_0m093ys8w16md_fm1mnxhtfnj8/T/encfs-tests-_8cR/create_remount.mnt): Resource busy -- try 'diskutil unmount'
integration/normal.t.pl ... 55/132 
#   Failed test 'encfs command returns 0'
#   at integration/normal.t.pl line 441.
cannot opendir for /var/folders/c6/lnc_0m093ys8w16md_fm1mnxhtfnj8/T/encfs-tests-_8cR/create_remount.mnt: No such file or directory at integration/normal.t.pl line 388.
cannot remove directory for /var/folders/c6/lnc_0m093ys8w16md_fm1mnxhtfnj8/T/encfs-tests-_8cR/create_remount.mnt: Resource busy at integration/normal.t.pl line 388.
cannot remove directory for /var/folders/c6/lnc_0m093ys8w16md_fm1mnxhtfnj8/T//encfs-tests-_8cR: Directory not empty at integration/normal.t.pl line 388.

#   Failed test 'working dir removed'
#   at integration/normal.t.pl line 389.

runTests: mode=paranoia sudo=off
integration/normal.t.pl ... 119/132 umount(/private/var/folders/c6/lnc_0m093ys8w16md_fm1mnxhtfnj8/T/encfs-tests-_0i4/create_remount.mnt): Resource busy -- try 'diskutil unmount'
integration/normal.t.pl ... 122/132 
#   Failed test 'encfs command returns 0'
#   at integration/normal.t.pl line 441.
integration/normal.t.pl ... 128/132 cannot opendir for /var/folders/c6/lnc_0m093ys8w16md_fm1mnxhtfnj8/T/encfs-tests-_0i4/create_remount.mnt: No such file or directory at integration/normal.t.pl line 388.
cannot remove directory for /var/folders/c6/lnc_0m093ys8w16md_fm1mnxhtfnj8/T/encfs-tests-_0i4/create_remount.mnt: Resource busy at integration/normal.t.pl line 388.
cannot remove directory for /var/folders/c6/lnc_0m093ys8w16md_fm1mnxhtfnj8/T//encfs-tests-_0i4: Directory not empty at integration/normal.t.pl line 388.

#   Failed test 'working dir removed'
#   at integration/normal.t.pl line 389.
# Looks like you planned 132 tests but ran 130.
# Looks like you failed 4 tests of 130 run.
integration/normal.t.pl ... Dubious, test returned 4 (wstat 1024, 0x400)
Failed 6/132 subtests 
	(less 2 skipped subtests: 124 okay)
integration/reverse.t.pl .. ok     

Test Summary Report
-------------------
integration/normal.t.pl (Wstat: 1024 Tests: 130 Failed: 4)
  Failed tests:  55, 63, 122, 130
  Non-zero exit status: 4
  Parse errors: Bad plan.  You planned 132 tests but ran 130.
Files=2, Tests=161, 25 wallclock secs ( 0.04 usr  0.01 sys + 21.93 cusr  1.17 csys = 23.15 CPU)
Result: FAIL
Failed 1/2 test programs. 4/161 subtests failed.
make[3]: *** [CMakeFiles/integration] Error 255
make[2]: *** [CMakeFiles/integration.dir/all] Error 2
make[1]: *** [CMakeFiles/integration.dir/rule] Error 2
make: *** [integration] Error 2
$ mount
/dev/disk1 on / (hfs, local, journaled)
devfs on /dev (devfs, local, nobrowse)
. . . . .
encfs@osxfuse1 on /private/var/folders/c6/lnc_0m093ys8w16md_fm1mnxhtfnj8/T/encfs-tests-_8cR/create_remount.mnt (osxfuse, local, nodev, nosuid, synchronous, mounted by ur20980)
encfs@osxfuse2 on /private/var/folders/c6/lnc_0m093ys8w16md_fm1mnxhtfnj8/T/encfs-tests-_0i4/create_remount.mnt (osxfuse, local, nodev, nosuid, synchronous, mounted by ur20980)
$
```

Now failure appears consistent on MacOS High Sierra 10.13.1 with Xcode-9.1:
```
. . . . .
[100%] Built target unittests
Running tests...
Test project /Users/uri/src/encfs/build
    Start 1: unit
1/1 Test #1: unit .............................   Passed    0.08 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   0.09 sec
Scanning dependencies of target integration
++ dirname ../integration.sh
+ cd ..
+ perl -I. -MTest::Harness -e '$$Test::Harness::debug=1; runtests @ARGV;' integration/normal.t.pl integration/reverse.t.pl
integration/normal.t.pl ... 
runTests: mode=standard sudo=off
integration/normal.t.pl ... 52/132 umount(/private/var/folders/pd/mxn5kp_55jg23x7jjd10gtwm0000gn/T/encfs-tests-oXMA/create_remount.mnt): Resource busy -- try 'diskutil unmount'

#   Failed test 'encfs command returns 0'
#   at integration/normal.t.pl line 441.
integration/normal.t.pl ... 61/132 cannot remove directory for /var/folders/pd/mxn5kp_55jg23x7jjd10gtwm0000gn/T/encfs-tests-oXMA/create_remount.mnt: Resource busy at integration/normal.t.pl line 388.
cannot remove directory for /var/folders/pd/mxn5kp_55jg23x7jjd10gtwm0000gn/T//encfs-tests-oXMA: Directory not empty at integration/normal.t.pl line 388.

#   Failed test 'working dir removed'
#   at integration/normal.t.pl line 389.

runTests: mode=paranoia sudo=off
integration/normal.t.pl ... 117/132 umount(/private/var/folders/pd/mxn5kp_55jg23x7jjd10gtwm0000gn/T/encfs-tests-1B8p/create_remount.mnt): Resource busy -- try 'diskutil unmount'
integration/normal.t.pl ... 122/132 
#   Failed test 'encfs command returns 0'
#   at integration/normal.t.pl line 441.
integration/normal.t.pl ... 128/132 cannot remove directory for /var/folders/pd/mxn5kp_55jg23x7jjd10gtwm0000gn/T/encfs-tests-1B8p/create_remount.mnt: Resource busy at integration/normal.t.pl line 388.
cannot remove directory for /var/folders/pd/mxn5kp_55jg23x7jjd10gtwm0000gn/T//encfs-tests-1B8p: Directory not empty at integration/normal.t.pl line 388.

#   Failed test 'working dir removed'
#   at integration/normal.t.pl line 389.
# Looks like you planned 132 tests but ran 130.
# Looks like you failed 4 tests of 130 run.
integration/normal.t.pl ... Dubious, test returned 4 (wstat 1024, 0x400)
Failed 6/132 subtests 
	(less 2 skipped subtests: 124 okay)
integration/reverse.t.pl .. ok     

Test Summary Report
-------------------
integration/normal.t.pl (Wstat: 1024 Tests: 130 Failed: 4)
  Failed tests:  55, 63, 122, 130
  Non-zero exit status: 4
  Parse errors: Bad plan.  You planned 132 tests but ran 130.
Files=2, Tests=161, 26 wallclock secs ( 0.06 usr  0.01 sys + 21.83 cusr  1.63 csys = 23.53 CPU)
Result: FAIL
Failed 1/2 test programs. 4/161 subtests failed.
make[3]: *** [CMakeFiles/integration] Error 255
make[2]: *** [CMakeFiles/integration.dir/all] Error 2
make[1]: *** [CMakeFiles/integration.dir/rule] Error 2
make: *** [integration] Error 2
```

@benrubson How dangerous is it to rely on this code? Is this failure fairly benign, or am I risking losing data?

--- a/encfs/FileUtils.cpp
+++ b/encfs/FileUtils.cpp
@@ -1686,6 +1686,15 @@ RootPtr initFS(EncFS_Context *ctx, const std::shared_ptr<EncFS_Opts> &opts) {
   return rootInfo;
 }
 
+void unmountFS(const char *mountPoint) {
+  // fuse_unmount succeeds and returns void
+  fuse_unmount(mountPoint, nullptr);
+#ifdef __APPLE__
+  // fuse_unmount does not work on Mac OS, see #428
+  unmount(mountPoint, MNT_FORCE);
+#endif
+}
+
 int remountFS(EncFS_Context *ctx) {
   VLOG(1) << "Attempting to reinitialize filesystem";
 
@@ -1706,13 +1715,8 @@ bool unmountFS(EncFS_Context *ctx) {
     ctx->setRoot(std::shared_ptr<DirNode>());
     return false;
   }
-// Time to unmount!
-  fuse_unmount(ctx->opts->mountPoint.c_str(), nullptr);
-#ifdef __APPLE__
-  // fuse_unmount does not work on Mac OS, see #428
-  unmount(ctx->opts->mountPoint.c_str(), MNT_FORCE);
-#endif
-  // fuse_unmount succeeds and returns void
+  // Time to unmount!
+  unmountFS(ctx->opts->mountPoint.c_str());
   RLOG(INFO) << "Filesystem inactive, unmounted: " << ctx->opts->mountPoint;
   return true;
 }

--- a/encfs/FileUtils.cpp
+++ b/encfs/FileUtils.cpp
@@ -35,6 +35,10 @@
 #include <fstream>
 #include <iostream>
 #include <list>
+#ifdef __APPLE__
+#include <sys/mount.h>
+#include <sys/param.h>
+#endif
 #include <sys/socket.h>
 #include <sys/stat.h>
 #include <sys/wait.h>
@@ -1703,10 +1707,10 @@ bool unmountFS(EncFS_Context *ctx) {
     return false;
   }
 // Time to unmount!
-#if FUSE_USE_VERSION < 30
   fuse_unmount(ctx->opts->mountPoint.c_str(), nullptr);
-#else
-  fuse_unmount(fuse_get_context()->fuse);
+#ifdef __APPLE__
+  // fuse_unmount does not work on Mac OS, see #428
+  unmount(ctx->opts->mountPoint.c_str(), MNT_FORCE);
 #endif
   // fuse_unmount succeeds and returns void
   RLOG(INFO) << "Filesystem inactive, unmounted: " << ctx->opts->mountPoint;

--- a/encfs/FileUtils.cpp
+++ b/encfs/FileUtils.cpp
@@ -222,10 +222,19 @@ ConfigType readConfig_load(ConfigInfo *nm, const char *path,
  * Try to locate the config file
  * Tries the most recent format first, then looks for older versions
  */
-ConfigType readConfig(const string &rootDir, EncFSConfig *config) {
+ConfigType readConfig(const string &rootDir, EncFSConfig *config, const string &cmdConfig) {
   ConfigInfo *nm = ConfigFileMapping;
   while (nm->fileName != nullptr) {
-    // allow environment variable to override default config path
+   // allow command line argument to override default config path 
+    if (!cmdConfig.empty()) {
+      if (!fileExists(cmdConfig.c_str())) {
+        RLOG(ERROR)
+            << "fatal: config file specified on command line does not exist: "
+            << cmdConfig;
+        exit(1);
+      }
+      return readConfig_load(nm, cmdConfig.c_str(), config);
+    }    // allow environment variable to override default config path
     if (nm->environmentOverride != nullptr) {
       char *envFile = getenv(nm->environmentOverride);
       if (envFile != nullptr) {
@@ -433,14 +442,18 @@ bool readV4Config(const char *configFile, EncFSConfig *config,
 }
 
 bool saveConfig(ConfigType type, const string &rootDir,
-                const EncFSConfig *config) {
+                const EncFSConfig *config, const string &cmdConfig) {
   bool ok = false;
 
   ConfigInfo *nm = ConfigFileMapping;
   while (nm->fileName != nullptr) {
     if (nm->type == type && (nm->saveFunc != nullptr)) {
       string path = rootDir + nm->fileName;
-      if (nm->environmentOverride != nullptr) {
+      if (!cmdConfig.empty()) {
+        // use command line argument if specified
+        path.assign(cmdConfig);
+      }
+      else if (nm->environmentOverride != nullptr) {
         // use environment file if specified..
         const char *envFile = getenv(nm->environmentOverride);
         if (envFile != nullptr) {
@@ -1199,7 +1212,7 @@ RootPtr createV6Config(EncFS_Context *ctx,
     return rootInfo;
   }
 
-  if (!saveConfig(Config_V6, rootDir, config.get())) {
+  if (!saveConfig(Config_V6, rootDir, config.get(), opts->config)) {
     return rootInfo;
   }
 
@@ -1559,7 +1572,7 @@ RootPtr initFS(EncFS_Context *ctx, const std::shared_ptr<EncFS_Opts> &opts) {
   RootPtr rootInfo;
   std::shared_ptr<EncFSConfig> config(new EncFSConfig);
 
-  if (readConfig(opts->rootDir, config.get()) != Config_None) {
+  if (readConfig(opts->rootDir, config.get(), opts->config) != Config_None) {
     if (config->blockMACBytes == 0 && opts->requireMac) {
       cout << _(
           "The configuration disabled MAC, but you passed --require-macs\n");

--- a/encfs/FileUtils.h
+++ b/encfs/FileUtils.h
@@ -136,6 +136,8 @@ class EncFS_Context;
 
 RootPtr initFS(EncFS_Context *ctx, const std::shared_ptr<EncFS_Opts> &opts);
 
+void unmountFS(const char *mountPoint);
+
 RootPtr createV6Config(EncFS_Context *ctx,
                        const std::shared_ptr<EncFS_Opts> &opts);
 

--- a/encfs/FileUtils.h
+++ b/encfs/FileUtils.h
@@ -76,6 +76,7 @@ struct EncFS_Opts {
   bool idleTracking;       // turn on idle monitoring of filesystem
   bool mountOnDemand;      // mounting on-demand
   bool delayMount;         // delay initial mount
+  bool unmount;            // unmount
 
   bool checkKey;     // check crypto key decoding
   bool forceDecode;  // force decode on MAC block failures
@@ -105,6 +106,7 @@ struct EncFS_Opts {
     idleTracking = false;
     mountOnDemand = false;
     delayMount = false;
+    unmount = false;
     checkKey = true;
     forceDecode = false;
     useStdin = false;

--- a/encfs/FileUtils.h
+++ b/encfs/FileUtils.h
@@ -98,6 +98,7 @@ struct EncFS_Opts {
   bool requireMac;  // Throw an error if MAC is disabled
 
   ConfigMode configMode;
+  std::string config;  // path to configuration file (or empty)
 
   EncFS_Opts() {
     createIfNotFound = true;
@@ -120,14 +121,14 @@ struct EncFS_Opts {
 /*
     Read existing config file.  Looks for any supported configuration version.
 */
-ConfigType readConfig(const std::string &rootDir, EncFSConfig *config);
+ConfigType readConfig(const std::string &rootDir, EncFSConfig *config, const std::string &cmdConfig);
 
 /*
     Save the configuration.  Saves back as the same configuration type as was
     read from.
 */
 bool saveConfig(ConfigType type, const std::string &rootdir,
-                const EncFSConfig *config);
+                const EncFSConfig *config, const std::string &cmdConfig);
 
 class EncFS_Context;
 

--- a/encfs/encfs.pod
+++ b/encfs/encfs.pod
@@ -16,7 +16,7 @@ encfs - mounts or creates an encrypted virtual filesystem
 
 =head1 SYNOPSIS
 
-B<encfs> [B<--version>] [B<-v>|B<--verbose>] [B<-t>|B<--syslogtag>] 
+B<encfs> [B<--version>] [B<-v>|B<--verbose>] [B<-c>|B<--config>] [B<-t>|B<--syslogtag>] 
 [B<-s>] [B<-f>] [B<--annotate>] [B<--standard>] [B<--paranoia>] 
 [B<--reverse>] [B<--reversewrite>] [B<--extpass=program>] [B<-S>|B<--stdinpass>] 
 [B<--anykey>] [B<--forcedecode>] [B<-require-macs>] 
@@ -47,6 +47,10 @@ may be an increasing number of choices.
 
 Shows B<EncFS> version.  Using B<--verbose> before B<--version> may display
 additional information.
+
+=item B<-c>, B<--config>
+
+Causes B<EncFS> to use the supplied file as the configuration file.
 
 =item B<-v>, B<--verbose>
 

--- a/encfs/encfs.pod
+++ b/encfs/encfs.pod
@@ -20,7 +20,7 @@ B<encfs> [B<--version>] [B<-v>|B<--verbose>] [B<-c>|B<--config>] [B<-t>|B<--sysl
 [B<-s>] [B<-f>] [B<--annotate>] [B<--standard>] [B<--paranoia>] 
 [B<--reverse>] [B<--reversewrite>] [B<--extpass=program>] [B<-S>|B<--stdinpass>] 
 [B<--anykey>] [B<--forcedecode>] [B<-require-macs>] 
-[B<-i MINUTES>|B<--idle=MINUTES>] [B<-m>|B<--ondemand>] [B<--delaymount>] 
+[B<-i MINUTES>|B<--idle=MINUTES>] [B<-m>|B<--ondemand>] [B<--delaymount>] [B<-u>|B<--unmount>] 
 [B<--public>] [B<--nocache>] [B<--no-default-flags>]
 [B<-o FUSE_OPTION>] [B<-d>|B<--fuse-debug>] [B<-H>|B<--fuse-help>] 
 I<rootdir> I<mountPoint> 
@@ -207,6 +207,10 @@ password.  If this succeeds, then the filesystem becomes available again.
 
 Do not mount the filesystem when encfs starts; instead, delay mounting until
 first use. This option only makes sense with B<--ondemand>.
+
+=item B<-u>, B<--unmount>
+
+Unmounts the specified I<mountPoint>.
 
 =item B<--public>
 

--- a/encfs/encfsctl.cpp
+++ b/encfs/encfsctl.cpp
@@ -164,7 +164,7 @@ static int showInfo(int argc, char **argv) {
   if (!checkDir(rootDir)) return EXIT_FAILURE;
 
   std::shared_ptr<EncFSConfig> config(new EncFSConfig);
-  ConfigType type = readConfig(rootDir, config.get());
+  ConfigType type = readConfig(rootDir, config.get(), "");
 
   // show information stored in config..
   switch (type) {
@@ -622,7 +622,7 @@ static int do_chpasswd(bool useStdin, bool annotate, bool checkOnly, int argc,
   if (!checkDir(rootDir)) return EXIT_FAILURE;
 
   EncFSConfig *config = new EncFSConfig;
-  ConfigType cfgType = readConfig(rootDir, config);
+  ConfigType cfgType = readConfig(rootDir, config, "");
 
   if (cfgType == Config_None) {
     cout << _("Unable to load or parse config file\n");
@@ -683,7 +683,7 @@ static int do_chpasswd(bool useStdin, bool annotate, bool checkOnly, int argc,
     config->assignKeyData(keyBuf, encodedKeySize);
     delete[] keyBuf;
 
-    if (saveConfig(cfgType, rootDir, config)) {
+    if (saveConfig(cfgType, rootDir, config, "")) {
       // password modified -- changes volume key of filesystem..
       cout << _("Volume Key successfully updated.\n");
       result = EXIT_SUCCESS;

--- a/encfs/main.cpp
+++ b/encfs/main.cpp
@@ -28,10 +28,6 @@
 #include <pthread.h>
 #include <sstream>
 #include <string>
-#ifdef __APPLE__
-#include <sys/mount.h>
-#include <sys/param.h>
-#endif
 #include <sys/stat.h>
 #include <sys/time.h>
 #include <unistd.h>
@@ -609,12 +605,7 @@ int main(int argc, char *argv[]) {
 
   // Let's unmount if requested
   if (encfsArgs->opts->unmount) {
-    fuse_unmount(encfsArgs->opts->mountPoint.c_str(), nullptr);
-#ifdef __APPLE__
-    // fuse_unmount does not work on Mac OS, see #428
-    unmount(encfsArgs->opts->mountPoint.c_str(), MNT_FORCE);
-#endif
-    // fuse_unmount succeeds and returns void
+    unmountFS(encfsArgs->opts->mountPoint.c_str());
     cout << "Filesystem unmounting: " << encfsArgs->opts->mountPoint << endl;
     return 0;
   }

--- a/encfs/main.cpp
+++ b/encfs/main.cpp
@@ -162,6 +162,8 @@ static void usage(const char *name) {
             "reverse encryption\n")
        << _("  --reversewrite\t\t"
             "reverse encryption with writes enabled\n")
+       << _("  -c, --config=path\t\t"
+            "specifies config file (overrides ENV variable)\n")
 
        // xgroup(usage)
        << _("  --extpass=program\tUse external program for password prompt\n"
@@ -255,6 +257,7 @@ static bool processArgs(int argc, char *argv[],
       {"standard", 0, nullptr, '1'},              // standard configuration
       {"paranoia", 0, nullptr, '2'},              // standard configuration
       {"require-macs", 0, nullptr, LONG_OPT_REQUIRE_MAC},  // require MACs
+      { "config", 1, nullptr, 'c'},               // command-line-supplied config location 
       {nullptr, 0, nullptr, 0}};
 
   while (true) {
@@ -269,8 +272,9 @@ static bool processArgs(int argc, char *argv[],
     // 'S' : password from stdin
     // 'o' : arguments meant for fuse
     // 't' : syslog tag
+    // 'c' : configuration file
     int res =
-        getopt_long(argc, argv, "HsSfvdmi:o:t:", long_options, &option_index);
+        getopt_long(argc, argv, "HsSfvdmi:o:t:c:", long_options, &option_index);
 
     if (res == -1) {
       break;
@@ -297,6 +301,11 @@ static bool processArgs(int argc, char *argv[],
         break;
       case LONG_OPT_REQUIRE_MAC:
         out->opts->requireMac = true;
+        break;
+      case 'c':
+        /* Take config file path from command 
+         * line instead of ENV variable */
+        out->opts->config.assign(optarg);
         break;
       case 'f':
         out->isDaemon = false;

--- a/encfs/main.cpp
+++ b/encfs/main.cpp
@@ -615,7 +615,7 @@ int main(int argc, char *argv[]) {
     unmount(encfsArgs->opts->mountPoint.c_str(), MNT_FORCE);
 #endif
     // fuse_unmount succeeds and returns void
-    cerr << "Filesystem unmounting: " << encfsArgs->opts->mountPoint << endl;
+    cout << "Filesystem unmounting: " << encfsArgs->opts->mountPoint << endl;
     return 0;
   }
 

--- a/encfs/main.cpp
+++ b/encfs/main.cpp
@@ -28,6 +28,10 @@
 #include <pthread.h>
 #include <sstream>
 #include <string>
+#ifdef __APPLE__
+#include <sys/mount.h>
+#include <sys/param.h>
+#endif
 #include <sys/stat.h>
 #include <sys/time.h>
 #include <unistd.h>
@@ -164,6 +168,8 @@ static void usage(const char *name) {
             "reverse encryption with writes enabled\n")
        << _("  -c, --config=path\t\t"
             "specifies config file (overrides ENV variable)\n")
+       << _("  -u, --unmount\t\t"
+            "unmounts specified mountPoint\n")
 
        // xgroup(usage)
        << _("  --extpass=program\tUse external program for password prompt\n"
@@ -219,6 +225,7 @@ static bool processArgs(int argc, char *argv[],
   out->opts->annotate = false;
   out->opts->reverseEncryption = false;
   out->opts->requireMac = false;
+  out->opts->unmount = false;
 
   bool useDefaultFlags = true;
 
@@ -257,7 +264,8 @@ static bool processArgs(int argc, char *argv[],
       {"standard", 0, nullptr, '1'},              // standard configuration
       {"paranoia", 0, nullptr, '2'},              // standard configuration
       {"require-macs", 0, nullptr, LONG_OPT_REQUIRE_MAC},  // require MACs
-      { "config", 1, nullptr, 'c'},               // command-line-supplied config location 
+      {"config", 1, nullptr, 'c'},                // command-line-supplied config location
+      {"unmount", 1, nullptr, 'u'},               // unmount
       {nullptr, 0, nullptr, 0}};
 
   while (true) {
@@ -273,8 +281,9 @@ static bool processArgs(int argc, char *argv[],
     // 'o' : arguments meant for fuse
     // 't' : syslog tag
     // 'c' : configuration file
+    // 'u' : unmount
     int res =
-        getopt_long(argc, argv, "HsSfvdmi:o:t:c:", long_options, &option_index);
+        getopt_long(argc, argv, "HsSfvdmi:o:t:c:u", long_options, &option_index);
 
     if (res == -1) {
       break;
@@ -306,6 +315,9 @@ static bool processArgs(int argc, char *argv[],
         /* Take config file path from command 
          * line instead of ENV variable */
         out->opts->config.assign(optarg);
+        break;
+      case 'u':
+        out->opts->unmount = true;
         break;
       case 'f':
         out->isDaemon = false;
@@ -385,7 +397,7 @@ static bool processArgs(int argc, char *argv[],
         break;
       case 'P':
         if (geteuid() != 0) {
-          RLOG(WARNING) << "option '--public' ignored for non-root user";
+          cerr << "option '--public' ignored for non-root user";
         } else {
           out->opts->ownerCreate = true;
           // add 'allow_other' option
@@ -416,13 +428,24 @@ static bool processArgs(int argc, char *argv[],
         // missing parameter for option..
         break;
       default:
-        RLOG(WARNING) << "getopt error: " << res;
+        cerr << "getopt error: " << res;
         break;
     }
   }
 
   if (!out->isThreaded) {
     PUSHARG("-s");
+  }
+
+  // for --unmount, we should have exactly 1 argument - the mount point
+  if (out->opts->unmount) {
+    if (optind + 1 == argc) {
+      out->opts->mountPoint = slashTerminate(argv[optind++]);
+      return true;
+    }
+    // no mount point specified
+    cerr << _("Expecting one argument, aborting.") << endl;
+    return false;
   }
 
   // we should have at least 2 arguments left over - the source directory and
@@ -434,7 +457,7 @@ static bool processArgs(int argc, char *argv[],
     out->opts->mountPoint = slashTerminate(argv[optind++]);
   } else {
     // no mount point specified
-    cerr << _("Missing one or more arguments, aborting.");
+    cerr << _("Missing one or more arguments, aborting.") << endl;
     return false;
   }
 
@@ -582,6 +605,18 @@ int main(int argc, char *argv[]) {
   if (argc == 1 || !processArgs(argc, argv, encfsArgs)) {
     usage(argv[0]);
     return EXIT_FAILURE;
+  }
+
+  // Let's unmount if requested
+  if (encfsArgs->opts->unmount) {
+    fuse_unmount(encfsArgs->opts->mountPoint.c_str(), nullptr);
+#ifdef __APPLE__
+    // fuse_unmount does not work on Mac OS, see #428
+    unmount(encfsArgs->opts->mountPoint.c_str(), MNT_FORCE);
+#endif
+    // fuse_unmount succeeds and returns void
+    cerr << "Filesystem unmounting: " << encfsArgs->opts->mountPoint << endl;
+    return 0;
   }
 
   encfs::initLogging(encfsArgs->isVerbose, encfsArgs->isDaemon);

--- a/integration/common.pl
+++ b/integration/common.pl
@@ -2,13 +2,7 @@
 # works on Linux AND OSX
 sub portable_unmount {
     my $crypt = shift;
-    my $fusermount = qx(which fusermount);
-    chomp($fusermount);
-    if(-f $fusermount) {
-        qx($fusermount -u "$crypt");
-    } else {
-        qx(umount "$crypt");
-    }
+    qx(./build/encfs -u "$crypt");
 }
 
 # Helper function

--- a/integration/common.pl
+++ b/integration/common.pl
@@ -2,7 +2,7 @@
 # works on Linux AND OSX
 sub portable_unmount {
     my $crypt = shift;
-    qx(./build/encfs -u "$crypt");
+    qx(./build/encfs -u "$crypt" >/dev/null);
 }
 
 # Helper function

--- a/integration/normal.t.pl
+++ b/integration/normal.t.pl
@@ -436,8 +436,9 @@ sub create_unmount_remount
     # Unmount
     portable_unmount($mnt);
 
-    # Mount again
-    system("./build/encfs --extpass=\"echo test\" $crypt $mnt 2>&1");
+    # Mount again, testing -c as the same time
+    rename("$crypt/.encfs6.xml", "$crypt/.encfs6_moved.xml");
+    system("./build/encfs -c $crypt/.encfs6_moved.xml --extpass=\"echo test\" $crypt $mnt 2>&1");
     ok( $? == 0, "encfs command returns 0") || return;
 
     # Check if content is still there

